File: app/src/main/java/moe/matsuri/nb4a/utils/JavaUtil.java
package moe.matsuri.nb4a.utils;

import android.annotation.SuppressLint;
import android.app.Application;
import android.os.Build;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.ToNumberPolicy;

import java.lang.reflect.Method;

import io.nekohasekai.sagernet.BuildConfig;
import kotlin.text.StringsKt;

public class JavaUtil {

    @SuppressLint("PrivateApi")
    public static String getProcessName() {
        if (Build.VERSION.SDK_INT >= 28)
            return Application.getProcessName();

        // Use the same technique as Application.getProcessName() on older Android versions.
        try {
            Class<?> activityThread = Class.forName("android.app.ActivityThread");
            Method getProcessName = activityThread.getDeclaredMethod("currentProcessName");
            return (String) getProcessName.invoke(null);
        } catch (Exception e) {
            return BuildConfig.APPLICATION_ID;
        }
    }

    public static boolean isNullOrBlank(String str) {
        return str == null || StringsKt.isBlank(str);
    }

    public static boolean isNotBlank(String str) {
        return !isNullOrBlank(str);
    }

    private static final char[] HEX_ARRAY = "0123456789abcdef".toCharArray();

    public static String bytesToHex(byte[] bytes) {
        char[] hexChars = new char[bytes.length * 2];
        for (int j = 0; j < bytes.length; j++) {
            int v = bytes[j] & 0xFF;
            hexChars[j * 2] = HEX_ARRAY[v >>> 4];
            hexChars[j * 2 + 1] = HEX_ARRAY[v & 0x0F];
        }
        return new String(hexChars);
    }

    public static boolean isEmpty(byte[] array) {
        return array == null || array.length == 0;
    }

    public static final Gson gson = new GsonBuilder()
            .setPrettyPrinting()
            .setNumberToNumberStrategy(ToNumberPolicy.LONG_OR_DOUBLE)
            .setObjectToNumberStrategy(ToNumberPolicy.LONG_OR_DOUBLE)
            .setLenient()
            .disableHtmlEscaping()
            .create();
}

File: app/src/main/java/moe/matsuri/nb4a/utils/JavaUtil.kt
package moe.matsuri.nb4a.utils

import android.annotation.SuppressLint
import android.app.Application
import android.os.Build
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.Strictness
import com.google.gson.ToNumberPolicy
import io.nekohasekai.sagernet.BuildConfig

/** Android/JVM helpers implemented in Kotlin. */
object JavaUtil {
    @SuppressLint("PrivateApi")
    @JvmStatic
    fun getProcessName(): String {
        if (Build.VERSION.SDK_INT >= 28) return Application.getProcessName()
        return runCatching {
            val activityThread = Class.forName("android.app.ActivityThread")
            val method = activityThread.getDeclaredMethod("currentProcessName")
            method.invoke(null) as String
        }.getOrDefault(BuildConfig.APPLICATION_ID)
    }

    @JvmStatic
    fun isNullOrBlank(value: String?): Boolean = value.isNullOrBlank()

    @JvmStatic
    fun isNotBlank(value: String?): Boolean = !value.isNullOrBlank()

    @JvmStatic
    fun isEmpty(array: ByteArray?): Boolean = array == null || array.isEmpty()

    @JvmField
    val gson: Gson = GsonBuilder()
        .setPrettyPrinting()
        .setNumberToNumberStrategy(ToNumberPolicy.LONG_OR_DOUBLE)
        .setObjectToNumberStrategy(ToNumberPolicy.LONG_OR_DOUBLE)
        .setStrictness(Strictness.LENIENT)
        .disableHtmlEscaping()
        .create()
}
